torch/csrc/jit/fuser/cuda/fused_kernel.cpp

```diff
@@ -0,0 +1,187 @@
+#include "THC/THCTensorRandom.h"
+#include "THC/THCGenerator.hpp"
+THCGenerator* THCRandom_getGenerator(THCState* state);
+
+#include "nvrtc.h"
+#include "cuda.h"
+#include "cuda_runtime.h"
+
+#include <stdexcept>
+#include <sstream>
+#include <tuple>
+#include <vector>
+#include <algorithm>
+#include <cmath>
+
+namespace torch { namespace jit { namespace fuser { namespace cuda {
+
+void checkCUDAVersion(
+  const cudaDeviceProp& prop) {
+  if ((prop.major >= 6 && CUDA_VERSION < 8000) ||
+      (prop.major >= 7 && CUDA_VERSION < 9000)) {
+    std::stringstream err_string;
+    err_string << "In CUDAFusedKernel, PyTorch compiled with insufficient CUDA version: "
+         << CUDA_VERSION << " for the current GPU device " << prop.name
+         << " with device capability " << prop.major << "." << prop.minor;
+    throw std::runtime_error(err_string.str());
+  }
+}
+
+static void getMajorMinor(const cudaDeviceProp& prop, int& major, int& minor) {
+  int nvrtc_major, nvrtc_minor;
+  TORCH_NVRTC_CHECK(nvrtcVersion(&nvrtc_major, &nvrtc_minor));
+
+  // Short-circuits if NVRTC version too low
+  JIT_ASSERT(nvrtc_major >= 6);
+
+  // Major and minor is determined by device properties and 
+  // possibly "downcompiled" to a lower (compatible) compute architecture
+  // based on the NVRTC version
+  major = prop.major;
+  minor = prop.minor;
+  if (nvrtc_major <= 7 && prop.major > 5) { // 7 supports 2-5.x
+    major = 5;
+    if (prop.major == 5) minor = prop.minor;
+    else minor = 0;
+  } else if (nvrtc_major <= 8 && prop.major > 6) { // 8 supports 2-6.x
+    major = 6;
+    if (prop.major == 6) minor = prop.minor;
+    else minor = 0;
+  } else if (nvrtc_major <= 9 && prop.major >= 7) { // 9 supports 3-7.2
+    major = 7;
+    if (prop.major == 7 && prop.minor <= 2) minor = prop.minor;
+    else minor = 0;
+  } else if (nvrtc_major <= 10 && prop.major >= 7) { // 10 supports 3-7.5
+    major = 7;
+    if (prop.major == 7 && prop.minor <= 5) minor = prop.minor;
+    else minor = 0;
+  }
+}
+
+// Compiles the specified kernel and stores the metadata required to run it
+FusedKernelCUDA::FusedKernelCUDA(
+  const int _device
+, const std::string& _name
+, const std::string& _code
+, const std::vector<TensorDesc> _input_desc
+, const std::vector<TensorDesc> _output_desc
+, const std::vector<PartitionDesc> _chunk_desc
+, const std::vector<PartitionDesc> _concat_desc
+, const bool _has_random)
+: FusedKernel{_name, _code, _input_desc, _output_desc, _chunk_desc, _concat_desc, _has_random} 
+, device_{_device} {
+  // Initializes driver's API context (if necessary)
+  CUcontext pctx = 0;
+  TORCH_CU_CHECK(cuCtxGetCurrent(&pctx));
+  if (!pctx) {
+     std::unique_lock<std::mutex> cudaFreeMutexLock(
+     *(THCCachingAllocator_getCudaFreeMutex()));
+     cudaFree(0);
+  }
+
+  // Note: hacked at::DeviceGuard since at::DeviceGuard was failing to work
+  // properly in some scenarios
+  int prior_device;
+  cudaGetDevice(&prior_device);
```

Comment:
using cudart functions without error checking is dangerous. Aren't there at:: functions for getDevice/setDevice?

Reply:
Of course, thanks for reminding me.